RecoTracker/LSTCore/interface/alpaka/Module.h

```diff
@@ -0,0 +1,372 @@
+    uint16_t* indicesOfEligibleT5Modules;
+    // To store different starting points for variable occupancy stuff
+    int* quintupletModuleIndices;
+    int* quintupletModuleOccupancy;
+    int* miniDoubletModuleIndices;
+    int* miniDoubletModuleOccupancy;
+    int* segmentModuleIndices;
+    int* segmentModuleOccupancy;
+    int* tripletModuleIndices;
+    int* tripletModuleOccupancy;
+
+    unsigned int* device_nTotalMDs;
+    unsigned int* device_nTotalSegs;
+    unsigned int* device_nTotalTrips;
+    unsigned int* device_nTotalQuints;
+
+    template <typename TBuff>
+    void setData(TBuff& objectRangesbuf) {
+      hitRanges = alpaka::getPtrNative(objectRangesbuf.hitRanges_buf);
+      hitRangesLower = alpaka::getPtrNative(objectRangesbuf.hitRangesLower_buf);
+      hitRangesUpper = alpaka::getPtrNative(objectRangesbuf.hitRangesUpper_buf);
+      hitRangesnLower = alpaka::getPtrNative(objectRangesbuf.hitRangesnLower_buf);
+      hitRangesnUpper = alpaka::getPtrNative(objectRangesbuf.hitRangesnUpper_buf);
+      mdRanges = alpaka::getPtrNative(objectRangesbuf.mdRanges_buf);
+      segmentRanges = alpaka::getPtrNative(objectRangesbuf.segmentRanges_buf);
+      trackletRanges = alpaka::getPtrNative(objectRangesbuf.trackletRanges_buf);
+      tripletRanges = alpaka::getPtrNative(objectRangesbuf.tripletRanges_buf);
+      trackCandidateRanges = alpaka::getPtrNative(objectRangesbuf.trackCandidateRanges_buf);
+      quintupletRanges = alpaka::getPtrNative(objectRangesbuf.quintupletRanges_buf);
+
+      nEligibleT5Modules = alpaka::getPtrNative(objectRangesbuf.nEligibleT5Modules_buf);
+      indicesOfEligibleT5Modules = alpaka::getPtrNative(objectRangesbuf.indicesOfEligibleT5Modules_buf);
+
+      quintupletModuleIndices = alpaka::getPtrNative(objectRangesbuf.quintupletModuleIndices_buf);
+      quintupletModuleOccupancy = alpaka::getPtrNative(objectRangesbuf.quintupletModuleOccupancy_buf);
+      miniDoubletModuleIndices = alpaka::getPtrNative(objectRangesbuf.miniDoubletModuleIndices_buf);
+      miniDoubletModuleOccupancy = alpaka::getPtrNative(objectRangesbuf.miniDoubletModuleOccupancy_buf);
+      segmentModuleIndices = alpaka::getPtrNative(objectRangesbuf.segmentModuleIndices_buf);
+      segmentModuleOccupancy = alpaka::getPtrNative(objectRangesbuf.segmentModuleOccupancy_buf);
+      tripletModuleIndices = alpaka::getPtrNative(objectRangesbuf.tripletModuleIndices_buf);
+      tripletModuleOccupancy = alpaka::getPtrNative(objectRangesbuf.tripletModuleOccupancy_buf);
+
+      device_nTotalMDs = alpaka::getPtrNative(objectRangesbuf.device_nTotalMDs_buf);
+      device_nTotalSegs = alpaka::getPtrNative(objectRangesbuf.device_nTotalSegs_buf);
+      device_nTotalTrips = alpaka::getPtrNative(objectRangesbuf.device_nTotalTrips_buf);
+      device_nTotalQuints = alpaka::getPtrNative(objectRangesbuf.device_nTotalQuints_buf);
+    }
+  };
+
+  template <typename TDev>
+  struct objectRangesBuffer : objectRanges {
```

Comment:
This inheritance looks suspicious.

Reply:
how so?
The same pattern is used for all buffers in LST. The "base" is the POD/SoA, while the buffer is device-templated.

Reply:
Just to add, this pattern was implemented when we migrated to Alpaka to fix compilation errors related to overloading kernel input on the CUDA backend.

Reply:
Generally speaking inheritance creates coupling between components in ways that increase maintenance cost, and should be limited in cases where dynamic polymorphism is used (despite it giving the feeling of being handy to "import base class functions into derived class").

I'd think separation of trivially copyable "View" (`objectRanges`) and the memory owning object (`objectRangesBuffer<TDev>`), that e.g. the `PortableCollection` + `DataFormats/SoATemplate` follow (and `std` with e.g. `string_view` and `span`), would be useful here too.

I haven't looked at all how exactly these get used in the present code (and won't have time until some time July), but I wanted to note in any case that this pattern "smells". Depending on various things it may be sufficient to address in medium-to-long term.

Reply:
After seeing how these are being used, my take continues to be that separating the owner type and the non-owning "View" type would be the best way to go.

Do you have any plans to use `DataFormats/SoATemplate` + `DataFormats/Portable` for the SoAs? They would allow to reduce boilerplate code, and the number of memory buffers (one buffer per SoA instead of one buffer per column/array).